File: pyScripts/plot_figure2.py
import numpy as np
import torch
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.gridspec import GridSpec
import pandas as pd

def load_model_data(model_path, sig_refs_path):
    """Load model and signature reference data"""
    model = torch.load(model_path, weights_only=False)
    sig_refs = torch.load(sig_refs_path, weights_only=False)
    
    # Convert tensors to numpy arrays
    sigs = sig_refs['signature_refs'].detach().cpu().numpy()
    phi = model['model_state_dict']['phi'].detach().cpu().numpy()
    lambda_ = model['model_state_dict']['lambda'].detach().cpu().numpy()
    kappa = model['model_state_dict']['kappa'].detach().cpu().numpy()
    
    return sigs, phi, lambda_, kappa

def center_phi(phi):
    """Center phi by temporal mean for each disease"""
    phi_centered = np.zeros_like(phi)
    for k in range(phi.shape[0]):
        for d in range(phi.shape[1]):
            phi_centered[k, d, :] = phi[k, d, :] - np.mean(phi[:, d, :], axis=0)
    return phi_centered

def plot_signature_disease_associations(phi_centered, time_points=[0, 5, 10, 15, 20], n_top_diseases=5):
    """Panel A: Show how signature-disease associations evolve over time"""
    fig, axes = plt.subplots(1, len(time_points), figsize=(15, 5))
    
    # Calculate overall phi average for sorting
    phi_avg = phi_centered.mean(axis=2)
    max_sig = np.argmax(np.abs(phi_avg), axis=0)
    disease_order = np.argsort(max_sig)
    
    for i, t in enumerate(time_points):
        # Get phi at time t
        phi_t = phi_centered[:, :, t]
        phi_t_sorted = phi_t[:, disease_order]
        
        # Create heatmap with centered values
        vmax = np.max(np.abs(phi_centered))
        im = axes[i].imshow(phi_t_sorted, aspect='auto', cmap='RdBu_r', 
                          vmin=-vmax, vmax=vmax)
        axes[i].set_title(f'Time {t}')
        
        if i == 0:
            axes[i].set_ylabel('Signature Index')
        if i == len(time_points)//2:
            axes[i].set_xlabel('Diseases (sorted by primary signature)')
    
    plt.colorbar(im, ax=axes, label='Phi (centered)')
    plt.suptitle('A: Signature-Disease Associations Over Time')
    return fig

def plot_disease_signature_relationships(phi_centered, selected_diseases, disease_names=None):
    """Panel B: Show probabilistic relationships between diseases and signatures"""
    if disease_names is None:
        disease_names = [f"Disease {i}" for i in range(phi_centered.shape[1])]
    
    fig, axes = plt.subplots(len(selected_diseases), 1, figsize=(10, 3*len(selected_diseases)))
    
    for i, disease_idx in enumerate(selected_diseases):
        # Plot centered phi values for each signature
        for k in range(phi_centered.shape[0]):
            axes[i].plot(phi_centered[k, disease_idx, :], 
                        label=f'Signature {k}')
        
        axes[i].set_title(f'B: {disease_names[disease_idx][:40]}...')
        axes[i].set_ylabel('Phi (centered)')
        axes[i].legend(bbox_to_anchor=(1.05, 1), loc='upper left', fontsize='small')
    
    axes[-1].set_xlabel('Time')
    return fig

def plot_population_signature_evolution(lambda_):
    """Panel C: Show how signature prevalence changes across age cohorts"""
    # Calculate population-level theta
    theta = np.exp(lambda_ - np.max(lambda_, axis=1, keepdims=True))
    theta = theta / np.sum(theta, axis=1, keepdims=True)
    
    fig, ax = plt.subplots(figsize=(10, 6))
    
    # Plot each signature's prevalence over time
    for k in range(theta.shape[0]):
        ax.plot(theta[k, :], label=f'Signature {k}')
    
    ax.set_xlabel('Age')
    ax.set_ylabel('Population Prevalence')
    ax.set_title('C: Population-Level Signature Evolution')
    ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
    return fig

def plot_signature_trajectories(phi_centered, selected_signatures, disease_names=None):
    """Panel D: Show complete trajectories for selected signatures"""
    if disease_names is None:
        disease_names = [f"Disease {i}" for i in range(phi_centered.shape[1])]
    
    n_signatures = len(selected_signatures)
    fig, axes = plt.subplots(n_signatures, 2, figsize=(15, 3*n_signatures))
    
    for i, sig_idx in enumerate(selected_signatures):
        # Plot signature's top diseases over time
        top_diseases = np.argsort(-np.abs(phi_centered[sig_idx].mean(axis=1)))[:5]
        for d in top_diseases:
            axes[i, 0].plot(phi_centered[sig_idx, d, :], 
                          label=f'{disease_names[d][:30]}...')
        axes[i, 0].set_title(f'D: Signature {sig_idx} Top Diseases')
        axes[i, 0].legend(bbox_to_anchor=(1.05, 1), loc='upper left', fontsize='small')
        
        # Plot heatmap of all diseases for this signature
        vmax = np.max(np.abs(phi_centered))
        im = axes[i, 1].imshow(phi_centered[sig_idx], aspect='auto', 
                             cmap='RdBu_r', vmin=-vmax, vmax=vmax)
        axes[i, 1].set_title(f'Signature {sig_idx} All Diseases')
        axes[i, 1].set_xlabel('Time')
        axes[i, 1].set_ylabel('Disease Index')
    
    plt.colorbar(im, ax=axes[:, 1], label='Phi (centered)')
    return fig

def main():
    # Load data
    model_path = "/Users/sarahurbut/Dropbox/resultshighamp/results/output_0_10000/model.pt"
    sig_refs_path = "/Users/sarahurbut/Dropbox/data_for_running/reference_trajectories.pt"
    sigs, phi, lambda_, kappa = load_model_data(model_path, sig_refs_path)
    
    # Center phi values
    phi_centered = center_phi(phi)
    
    # Create figure with 4 panels
    fig = plt.figure(figsize=(20, 15))
    gs = GridSpec(2, 2, figure=fig)
    
    # Panel A
    ax1 = fig.add_subplot(gs[0, 0])
    plot_signature_disease_associations(phi_centered)
    
    # Panel B
    ax2 = fig.add_subplot(gs[0, 1])
    plot_disease_signature_relationships(phi_centered, selected_diseases=[0, 1, 2])
    
    # Panel C
    ax3 = fig.add_subplot(gs[1, 0])
    plot_population_signature_evolution(lambda_)
    
    # Panel D
    ax4 = fig.add_subplot(gs[1, 1])
    plot_signature_trajectories(phi_centered, selected_signatures=[0, 1, 2])
    
    plt.tight_layout()
    plt.savefig('figure2.png', dpi=300, bbox_inches='tight')

if __name__ == "__main__":
    main() 

File: pyScripts/plot_figure2.R
library(tidyverse)
library(ggplot2)
library(patchwork)
library(viridis)
library(ggridges)
library(gridExtra)

# Load and process model data
load_model_data <- function(model_path, sig_refs_path) {
  model <- torch::load(model_path, weights_only = FALSE)
  sig_refs <- torch::load(sig_refs_path, weights_only = FALSE)
  
  # Convert tensors to R objects
  sigs <- as.array(sig_refs$signature_refs$detach()$cpu()$numpy())
  phi <- as.array(model$model_state_dict$phi$detach()$cpu()$numpy())
  lambda_ <- as.array(model$model_state_dict$lambda$detach()$cpu()$numpy())
  kappa <- as.array(model$model_state_dict$kappa$detach()$cpu()$numpy())
  
  return(list(sigs = sigs, phi = phi, lambda_ = lambda_, kappa = kappa))
}

# Helper function for softmax
softmax_by_k <- function(x) {
  # Apply softmax along K dimension (dimension 2)
  exp_x <- exp(x)
  sweep(exp_x, c(1,3), apply(exp_x, c(1,3), sum), "/")
}

# Calculate signature-specific probabilities
get_signature_probs <- function(phi, lambda_, kappa) {
  # Convert phi to probabilities using sigmoid
  phi_prob <- 1 / (1 + exp(-phi))
  
  # Calculate population-level theta using softmax
  theta <- softmax_by_k(lambda_)
  
  # Calculate average theta across individuals
  avg_theta <- apply(theta, c(2,3), mean)
  
  # Calculate signature-specific probabilities
  sig_probs <- array(0, dim = c(dim(phi)[1], dim(phi)[2], dim(phi)[3]))
  for(t in 1:dim(phi)[3]) {
    for(k in 1:dim(phi)[1]) {
      sig_probs[k,,t] <- phi_prob[k,,t] * avg_theta[k,t] * kappa
    }
  }
  
  return(sig_probs)
}

# Panel A: Show age effect and signature contributions
plot_age_signature_effects <- function(phi, lambda_, kappa, selected_diseases) {
  # Get signature-specific probabilities
  sig_probs <- get_signature_probs(phi, lambda_, kappa)
  
  # Calculate total probability and dominant signature
  total_probs <- apply(sig_probs, c(2,3), sum)
  dominant_sig <- apply(sig_probs, c(2,3), which.max)
  
  # Create data frame for plotting
  df <- data.frame(
    time = rep(1:dim(phi)[3], length(selected_diseases)),
    disease = rep(selected_diseases, each = dim(phi)[3]),
    total_prob = as.vector(total_probs[selected_diseases,]),
    dominant_sig = as.vector(dominant_sig[selected_diseases,])
  )
  
  p <- ggplot(df, aes(x = time, y = total_prob)) +
    geom_line(aes(color = factor(dominant_sig)), size = 1) +
    geom_ribbon(aes(ymin = 0, ymax = total_prob, fill = factor(dominant_sig)), alpha = 0.2) +
    scale_color_viridis_d() +
    scale_fill_viridis_d() +
    facet_wrap(~disease, scales = "free_y", nrow = 1) +
    labs(title = "A: Population-Level Disease Probabilities and Dominant Signatures",
         x = "Age",
         y = "Disease Probability",
         color = "Dominant Signature",
         fill = "Dominant Signature") +
    theme_minimal() +
    theme(legend.position = "right")
  
  return(p)
}

# Panel B: Show signature contributions
plot_signature_contributions <- function(phi, lambda_, kappa, selected_diseases) {
  # Get signature-specific probabilities
  sig_probs <- get_signature_probs(phi, lambda_, kappa)
  
  # Create data frame for plotting
  df <- data.frame()
  for(d in selected_diseases) {
    for(k in 1:dim(phi)[1]) {
      df <- rbind(df, data.frame(
        time = 1:dim(phi)[3],
        disease = d,
        signature = k,
        probability = sig_probs[k,d,]
      ))
    }
  }
  
  p <- ggplot(df, aes(x = time, y = probability, fill = factor(signature))) +
    geom_area(position = "stack") +
    scale_fill_viridis_d() +
    facet_wrap(~disease, scales = "free_y", nrow = 1) +
    labs(title = "B: Population-Level Signature Contributions",
         x = "Age",
         y = "Probability",
         fill = "Signature") +
    theme_minimal() +
    theme(legend.position = "right")
  
  return(p)
}

# Panel C: Show signature evolution
plot_signature_evolution <- function(phi, lambda_) {
  # Calculate signature strengths over time
  theta <- softmax_by_k(lambda_)
  
  # Calculate average signature strength
  avg_theta <- apply(theta, c(2,3), mean)
  
  # Create data frame for plotting
  df <- data.frame(
    signature = rep(1:dim(avg_theta)[1], dim(avg_theta)[2]),
    time = rep(1:dim(avg_theta)[2], each = dim(avg_theta)[1]),
    strength = as.vector(avg_theta)
  )
  
  p <- ggplot(df, aes(x = time, y = strength, fill = factor(signature))) +
    geom_area(position = "stack") +
    scale_fill_viridis_d() +
    labs(title = "C: Signature Evolution Over Time",
         x = "Age",
         y = "Signature Strength",
         fill = "Signature") +
    theme_minimal() +
    theme(legend.position = "right")
  
  return(p)
}

# Panel D: Show disease clusters
plot_disease_clusters <- function(phi, time_points = c(1, 10, 20)) {
  # Calculate disease correlations at different time points
  phi_prob <- 1 / (1 + exp(-phi))
  
  # Create data frame for plotting
  df <- data.frame()
  for(t in time_points) {
    cor_matrix <- cor(t(phi_prob[,,t]))
    for(i in 1:nrow(cor_matrix)) {
      for(j in 1:ncol(cor_matrix)) {
        if(i < j) {
          df <- rbind(df, data.frame(
            time = t,
            disease1 = i,
            disease2 = j,
            correlation = cor_matrix[i,j]
          ))
        }
      }
    }
  }
  
  p <- ggplot(df, aes(x = factor(disease1), y = factor(disease2), fill = correlation)) +
    geom_tile() +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", 
                        midpoint = 0, limits = c(-1, 1)) +
    facet_wrap(~time, nrow = 1) +
    labs(title = "D: Disease Correlations Over Time",
         x = "Disease",
         y = "Disease",
         fill = "Correlation") +
    theme_minimal() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_blank(),
          legend.position = "right")
  
  return(p)
}

# Plot top diseases for each signature
plot_top_diseases <- function(phi, selected_sigs = c(1,2,3)) {
  phi_prob <- 1 / (1 + exp(-phi))
  plots <- list()
  
  for(sig in selected_sigs) {
    # Get top diseases for this signature
    avg_phi <- apply(phi_prob[sig,,], 1, mean)
    top_diseases <- order(avg_phi, decreasing = TRUE)[1:5]
    
    df <- data.frame(
      time = rep(1:dim(phi)[3], length(top_diseases)),
      disease = factor(rep(top_diseases, each = dim(phi)[3])),
      value = as.vector(t(phi_prob[sig,top_diseases,]))
    )
    
    p <- ggplot(df, aes(x = time, y = value, color = disease)) +
      geom_line() +
      theme_minimal() +
      labs(title = paste("Signature", sig),
           x = "Age",
           y = "Probability") +
      theme(legend.position = "right")
    
    plots[[length(plots) + 1]] <- p
  }
  
  return(plots)
}

# Plot cluster correspondence
plot_cluster_correspondence <- function(phi, time_points = c(10, 30, 50)) {
  phi_prob <- 1 / (1 + exp(-phi))
  plots <- list()
  
  for(t in time_points) {
    cor_mat <- cor(t(phi_prob[,,t]))
    df <- expand.grid(x = 1:nrow(cor_mat), y = 1:ncol(cor_mat))
    df$correlation <- as.vector(cor_mat)
    
    p <- ggplot(df, aes(x = x, y = y, fill = correlation)) +
      geom_tile() +
      scale_fill_gradient2(low = "blue", mid = "white", high = "red",
                          midpoint = 0, limits = c(-1, 1)) +
      labs(title = paste("Age", t)) +
      theme_minimal() +
      theme(axis.text = element_blank())
    
    plots[[length(plots) + 1]] <- p
  }
  
  return(plots)
}

# Plot signature prevalence
plot_signature_prevalence <- function(lambda_) {
  theta <- softmax_by_k(lambda_)
  avg_theta <- apply(theta, c(2,3), mean)
  
  df <- data.frame(
    time = rep(1:dim(avg_theta)[2], dim(avg_theta)[1]),
    signature = factor(rep(1:dim(avg_theta)[1], each = dim(avg_theta)[2])),
    prevalence = as.vector(t(avg_theta))
  )
  
  p <- ggplot(df, aes(x = time, y = prevalence, color = signature)) +
    geom_line() +
    theme_minimal() +
    labs(title = "Signature Prevalence Across Age",
         x = "Age",
         y = "Prevalence") +
    theme(legend.position = "right")
  
  return(p)
}

# Main function
main <- function() {
  # Load data
  model_path <- "/Users/sarahurbut/Dropbox/resultshighamp/results/output_0_10000/model.pt"
  sig_refs_path <- "/Users/sarahurbut/Dropbox/data_for_running/reference_trajectories.pt"
  data <- load_model_data(model_path, sig_refs_path)
  
  # Create all plots
  top_disease_plots <- plot_top_diseases(data$phi)
  cluster_plots <- plot_cluster_correspondence(data$phi)
  prev_plot <- plot_signature_prevalence(data$lambda_)
  
  # Arrange plots using patchwork
  layout <- "
  AAABBB
  CCCDDD
  EEEFFF
  GGGGGG
  "
  
  final_plot <- (top_disease_plots[[1]] + top_disease_plots[[2]] +
                 top_disease_plots[[3]] + cluster_plots[[1]] +
                 cluster_plots[[2]] + cluster_plots[[3]] +
                 prev_plot) +
    plot_layout(design = layout) &
    theme(plot.margin = margin(5, 5, 5, 5))
  
  # Save plot
  ggsave("figure2.png", final_plot, width = 15, height = 12, dpi = 300)
}

# Run main function
main() 